dace/libraries/blas/nodes/gemm.py

```diff
@@ -0,0 +1,244 @@
+            elif shape_c == [1, N]:
+                memlet_idx = '0, __i1'
+            elif shape_c == [M, 1]:
+                memlet_idx = '__i0, 0'
+            elif shape_c == [N,]:
+                memlet_idx = '__i1'
+            else:
+                raise ValueError("Could not broadcast input _c to ({}, {})".format(M, N))
+                
+
+            # addition map
+            state.add_mapped_tasklet(
+                "_Add_",
+                {"__i%d" % i: "0:%s" % s
+                 for i, s in enumerate([M, N])}, {
+                     "__c": dace.Memlet.simple("_c", memlet_idx),
+                     "__tmp": dace.Memlet.simple(mul_out, "__i0, __i1"),
+                 },
+                add_program, {"__y": dace.Memlet.simple("_y", "__i0, __i1")},
+                external_edges=True,
+                input_nodes={mul_out: access_tmp})
+
+            sdfg.parent = parent_sdfg
+        sdfg.parent_sdfg = parent_sdfg
+
+        return sdfg
+
+    @staticmethod
+    def expansion(node, state, sdfg):
+        node.validate(sdfg, state)
+        if node.dtype is None:
+            raise ValueError("Data type must be set to expand " + str(node) +
+                             ".")
+        return ExpandGemmPure.make_sdfg(node, state, sdfg)
+
+
+@dace.library.node
+class Gemm(dace.graph.nodes.LibraryNode):
+    """Executes alpha * (A @ B) + beta * C. Broadcasting C is NOT supported.
+    """
+
+    # Global properties
+    implementations = {"pure": ExpandGemmPure}
+    default_implementation = None
+
+    # Object fields
+    dtype = dace.properties.TypeClassProperty(allow_none=True)
+    transA = Property(dtype=bool,
+                      desc="Whether to transpose A before multiplying")
+    transB = Property(dtype=bool,
+                      desc="Whether to transpose B before multiplying")
+    alpha = Property(
+        dtype=float,
+        desc="A scalar which will be multiplied with A @ B before adding C")
+    beta = Property(
+        dtype=float,
+        desc="A scalar which will be multiplied with C before adding C")
+
+    def __init__(self,
+                 name,
+                 dtype=None,
+                 location=None,
+                 transA=False,
+                 transB=False,
+                 alpha=1.0,
+                 beta=0.0):
+        super().__init__(name,
+                         location=location,
+                         inputs={"_a", "_b", "_c"},
+                         outputs={"_y"})
+        self.dtype = dtype
+        self.transA = transA
+        self.transB = transB
+        self.alpha = alpha
+        self.beta = beta
+
+    def validate(self, sdfg, state):
+        in_edges = state.in_edges(self)
+        if len(in_edges) != 3:
+            raise ValueError("Expected exactly three inputs to gemm")
+        for _, _, _, dst_conn, memlet in state.in_edges(self):
+            if dst_conn == "_a":
+                a_size = memlet.subset.size()
```

Comment:
Here (and in the validation), the MatMul node squeezes the shapes. I removed this so that broadcasting works.

Reply:
MatMul squeezes the shapes to do another kind of broadcasting in certain cases. Here it is ok because GEMM must explicitly work on two 2D arrays.